R/manual_layer_dag.R

```diff
@@ -0,0 +1,192 @@
+  #   d <- delayed(function(...) { sum(...) }, args=list(b,c), display_name='d', do_local=FALSE)
+  # has DAG like
+  #    a
+  #   / \
+  #  v   v
+  #  b   c
+  #   \ /
+  #    v
+  #    d
+  #
+  # What we do here is locate the initial and terminal nodes.
+  populate = function(.self, terminal_node) {
+    if (!is(terminal_node, "Node")) {
+      stop("Terminal node must be an instance of the Node class; got ", class(terminal_node))
+    }
+    .self$all_nodes <- list()
+    .self$initial_nodes <- list()
+    .self$populate_aux(terminal_node)
+    if (length(.self$all_nodes) == 0) {
+      stop("Task graph is empty")
+    }
+    if (length(.self$initial_nodes) == 0) {
+      stop("Task graph has a cycle")
+    }
+
+    .self$terminal_nodes <- list(terminal_node)
+
+    # This makes our print output non-frustrating for the user to read
+    .self$sort_nodes_topologically()
+  },
+
+  # This is a recursive helper function for populate().
+  populate_aux = function(.self, node) {
+    if (!is.null(.self$all_nodes[[node$id]])) {
+      # In R, just "return" does not actually ... return. ;)
+      return()
+    }
+
+    .self$all_nodes[[node$id]] <- node
+
+    if (node$is_initial()) {
+      .self$initial_nodes[[node$id]] <- node
+    }
+
+    for (arg in node$args) {
+      if (is(arg, "Node")) {
+        .self$populate_aux(arg)
+      }
+    }
+  },
+
+  # This is important so our status display is non-frustrating for the user.
+  # This assumes a cyclicity check has already been done.
+  # https://en.wikipedia.org/wiki/Topological_sorting Technically this is a
+  # reverse topological sort -- terminals appear first, then the ones that
+  # depend on them. This is intuitive for the user (and this topological sort
+  # is *only* for display to the user) since we normally write 'z <- f(x,y)'
+  # rather than 'f(x,y) -> z' -- so it makes sense to show z, then x and y.
+  sort_nodes_topologically = function(.self) {
+    if (length(.self$terminal_nodes) == 0) {
+      stop("terminal nodes must be identified before topological sort")
+    }
+    unsorted_nodes <- .self$all_nodes
+    sorted_nodes <- list()
+    terminal_nodes <- .self$terminal_nodes # this is a copy of a list of reference classes
+
+    while (length(terminal_nodes) > 0) {
+      node <- terminal_nodes[[1]]
+      terminal_nodes[[1]] <- NULL
+
+      sorted_nodes[[node$id]] <- node
+      for (arg in node$args) {
+        if (is(arg, "Node")) {
+          terminal_nodes[[arg$id]] <- arg
+        }
+      }
+    }
+
+    .self$all_nodes <- sorted_nodes
+  },
+
+  # ================================================================
+  # COMPUTE
+
+  # This is a synchronous wait.  If more inspection is desired, then one can do
+  # dag <- terminal_node$make_dag(), then dag$poll() and show(dag) / str(dag)
+  # repeatedly in order better to visualize the flow of computation through the
+  # graph.
+  compute = function(.self, timeout_seconds=120, verbose=FALSE) {
```

Comment:
My feeling is that it would be best for the default to be to wait indefinitely, since many computations may take more than 2 minutes.

Reply:
I set it to 10 million which is 3 years, ish ...

Reply:
done, thanks @thetorpedodog ! :)